### ISSUE TITLE
LG-304: Time based log rollover for Liberty

### ISSUE THREAD

**Issue by donbourne** [OpenLiberty/open-liberty#16067]
Provide time based rollover for message and trace log files in Liberty.

Suggested configuration:

```
<logging rolloverInterval="1d" rolloverStartTime="18:00"/>
```

rollover of messages.log and trace.log will occur on a schedule when rolloverInterval and/or rolloverStartTime are specified.

- rolloverInterval
  - will use duration type.
  - is optional - if rolloverStartTime is specified but rolloverInterval is not, rolloverInterval is assumed to be 1d

- rolloverStartTime
  - hh:mm format
    - both hours and minutes must be specified. 
    - padding zeroes must be specified for hour or minute numbers < 10
    - always uses 24 hour clock (as required by ISO-8601)
    - 00:00 is the only way to refer to midnight (24:00 is invalid)
    - no decimal places allowed (whole numbers only)
  - is optional - if rolloverInterval is specified but rolloverStartTime is not, rolloverStartTime is assumed to be 00:00

- when server starts, next rollover time is calculated as follows:
  - if rolloverStartTime is set:
    - if current time is < rolloverStartTime, first rollover will be at rolloverStartTime minus rolloverInterval x n as appropriate
    - else first rollover will be at rolloverStartTime + rolloverInterval x n as appropriate
  - else first rollover will be at midnight (at start of the day the server started) + rolloverInterval x n as appropriate
  - subsequent rollovers should occur every rolloverInterval

see RFE at https://www.ibm.com/developerworks/rfe/execute?use_case=viewChangeRequest&CR_ID=83861

**Comment by ellen-lau** [comment]
UFO: https://ibm.box.com/s/9oslyr5g7ey91gffszyz54cg3y0wk843

**Comment by yeekangc** [comment]
Feedback/Notes from UFO Review:
- Chart 12: Clarify when maxFiles/maxFileSize is set, what will happen if rollover attributes are set too; 
- Chart 13:
    - In the current UFO, rollover is assumed to start on the same day as current time;
    - What if the interval doesn't divide cleanly?  Will always work with whole numbers;
    - Clarify and find a better way to explain the behaviours in plain words to users.
- Chart 27: Is there consideration to apply the settings to the audit logs?  Does not apply to audit logs at the moment as audit logs are configured separately.  Clarify this for process logs and consider archiving of audit logs separately.

**Comment by ellen-lau** [comment]
Clarification on Feedback Points:

1. Clarify when maxFileSize is set, what will happen if rollover attributes are set too:

> This functionality will follow the current tWAS functionality, where a log rotation for maxFileSize will take precedence over a log rotation for rolloverStartTime/rolloverDuration. So if a log's maxFileSize is reached before the rolloverDuration is up, the log will be rolled for maxFileSize.

2. rolloverStartTime/rolloverInterval behaviour

> We've settled using the proposed behaviour in the UFO (just need to find a clearer way to explain the functionality).

3. Is there consideration to apply the settings to the audit logs?

> This will not apply to the audit logs at the moment, since audit logs are configured separately. If in the future, a customer requests this functionality for audit logs as well, then a separate epic will be opened for it.

**Comment by ellen-lau** [comment]
From an offline discussion: 

For the z/OS platform, this epic should have no effect on the z/OS MSGLOG DD. It looks like there is feature in JES2 that provides a something similar (can spin off a new MSGLOG DD on a timed basis), so there won't be anything done for MSGLOG, since the capability is already present.

**Comment by donbourne** [comment]
Discussion from 2021-11-25, 4:00 PM with @NottyCode 

- Aim to cover time based rollover for access logs in this epic (discuss with @pnicolucci first to make sure he agrees with approach)

---

Some thoughts on how we could cover time based rollover for access logs:

https://www.ibm.com/docs/en/was-liberty/base?topic=configuration-httpaccesslogging
add time based rollover attributes alongside maxFileSize / maxFiles.  would need to cover in 2 places:
- "anonymous" access logging config within scope of 1 endpoint
```
httpEndpoint
    accessLogging
```  
    
- access logging config with an id - shareable across HttpEndpoints 
```
httpAccessLogging
```

**Comment by donbourne** [comment]
**Serviceability Approval Comment** - Please answer the following questions for serviceability approval:

1.  UFO -- does the UFO identify the most likely problems customers will see and identify how the feature will enable them to diagnose and solve those problems without resorting to raising a PMR?  Have these issues been addressed in the implementation?

2.  Test and Demo -- As part of the serviceability process we're asking feature teams to test and analyze common problem paths for serviceability and demo those problem paths to someone not involved in the development of the feature (eg. L2, test team, or another development team).  
a) What problem paths were tested and demonstrated? 
b) Who did you demo to?
c) Do the people you demo'd to agree that the serviceability of the demonstrated problem scenarios is sufficient to avoid PMRs for any problems customers are likely to encounter, or that L2 should be able to quickly address those problems without need to engage L3?

3.  SVT -- SVT team is often the first team to try new features and often encounters problems setting up and using them.  Note that we're not expecting SVT to do full serviceability testing -- just to sign-off on the serviceability of the problem paths they encountered.
a) Who conducted SVT tests for this feature?
b) Do they agree that the serviceability of the problems they encountered is sufficient to avoid PMRs, or that L2 should be able to quickly address those problems without need to engage L3?  

4.  Which L2 / L3 queues will handle PMRs for this feature?  Ensure they are present in the contact reference file and in the queue contact summary, and that the respective L2/L3 teams know they are supporting it.  Ask Don Bourne if you need links or more info.

5.  Does this feature add any new metrics or emit any new JSON events?  If yes, have you updated the JMX metrics reference list / Metrics reference list / JSON log events reference list in the Open Liberty docs?

**Comment by ellen-lau** [comment]
Beta Blog here: #21259

**Comment by skasund** [comment]
@ellen-lau   L2 has requested STE slides for this feature.  The STE template can be found at the links below.  You can use either one to create the education.  

Slide Template:  https://ibm.box.com/s/1an42g7zdgmaj84w7dft0indqfgi8ffm

Github Template:  https://pages.github.ibm.com/WASL3/site/STE/about

Please upload the completed slides to the same 'STE Archive' BOX folder and let me know they're complete or provide me the Github link.   Thanks!

**Comment by ellen-lau** [comment]
@OpenLiberty/accessibility-approvers  N/A No accessibility concerns.

**Comment by ellen-lau** [comment]
@skasund, here is the link to the completed STE slides: https://ibm.box.com/s/0wsp64crgstpgtad1k4a4x8zl546nce1

I wasn't able to get to the 'STE Archive' BOX folder, since the template above was just a link -- will this shared link suffice?

**Comment by gnadell** [comment]
@ellen-lau , Thanks for the STE slides. Thanh Giang and I have shared the slides with L2 Team.

**Comment by ellen-lau** [comment]
From a discussion with @hanczaryk, it's been agreed that no explicit SVT is required for this feature since the FAT is sufficient. Brian can't log in to GH, so I am adding the `focalApproved:svt` label accordingly.

**Comment by gnadell** [comment]
Hi Ellen,
Please also include @Thanh ***@***.***> in all the future communications.


Thanks,
Geeta Nadella,
IBM Automation: WebSphere Application Server (WASCET) L2 Support
1-7203950423 (Office)
26760423 (Jabber ITN)
***@***.***(email)
@Geeta Nadella(slack)
[cid:95d91142-f265-4429-8ece-869a0a9dbb17][cid:af091e87-4447-40fe-abf7-81fd3bdf6d41][cid:92ed50f7-3828-4895-8532-35dc6bd0474a][cid:87b8ed46-13f4-466e-a2b9-fd88b82c75f9][cid:3fa87a1a-7476-4755-90a9-80e22c6f119d][cid:aa16ce06-8aa5-4fd3-809c-3c771d7be847]
________________________________
From: ellen-lau ***@***.***>
Sent: Monday, July 11, 2022 11:15 AM
To: OpenLiberty/open-liberty ***@***.***>
Cc: Geeta Nadella ***@***.***>; Comment ***@***.***>
Subject: [EXTERNAL] Re: [OpenLiberty/open-liberty] LG-304: Time based log rollover for Liberty (#16067)

From a discussion with @hanczaryk, it's been agreed that no explicit SVT is required for this feature since the FAT is sufficient. Brian can't log in to GH, so I am adding the focalApproved:svt label accordingly. — Reply to this email
ZjQcmQRYFpfptBannerStart
This Message Is From an External Sender
This message came from outside your organization.

ZjQcmQRYFpfptBannerEnd

From a discussion with @hanczaryk<https://github.com/hanczaryk>, it's been agreed that no explicit SVT is required for this feature since the FAT is sufficient. Brian can't log in to GH, so I am adding the focalApproved:svt label accordingly.

**Comment by ellen-lau** [comment]
@OpenLiberty/performance-approvers: We don't expect a performance hit in this feature, may I get your approval on performance? Thanks!

**Comment by donbourne** [comment]
**Serviceability Approval Comment** - Please answer the following questions for serviceability approval:

1.  UFO -- does the UFO identify the most likely problems customers will see and identify how the feature will enable them to diagnose and solve those problems without resorting to raising a PMR?  Have these issues been addressed in the implementation?

2.  Test and Demo -- As part of the serviceability process we're asking feature teams to test and analyze common problem paths for serviceability and demo those problem paths to someone not involved in the development of the feature (eg. L2, test team, or another development team).  
a) What problem paths were tested and demonstrated? 
b) Who did you demo to?
c) Do the people you demo'd to agree that the serviceability of the demonstrated problem scenarios is sufficient to avoid PMRs for any problems customers are likely to encounter, or that L2 should be able to quickly address those problems without need to engage L3?

3.  SVT -- SVT team is often the first team to try new features and often encounters problems setting up and using them.  Note that we're not expecting SVT to do full serviceability testing -- just to sign-off on the serviceability of the problem paths they encountered.
a) Who conducted SVT tests for this feature?
b) Do they agree that the serviceability of the problems they encountered is sufficient to avoid PMRs, or that L2 should be able to quickly address those problems without need to engage L3?  

4.  Which L2 / L3 queues will handle PMRs for this feature?  Ensure they are present in the contact reference file and in the queue contact summary, and that the respective L2/L3 teams know they are supporting it.  Ask Don Bourne if you need links or more info.

5.  Does this feature add any new metrics or emit any new JSON events?  If yes, have you updated the JMX metrics reference list / Metrics reference list / JSON log events reference list in the Open Liberty docs?

**Comment by ellen-lau** [comment]
Feature Test Summary: https://github.com/OpenLiberty/open-liberty/issues/21745

**Comment by ellen-lau** [comment]
ID Issue: https://github.com/OpenLiberty/docs/issues/5637

**Comment by ellen-lau** [comment]
Serviceability Approval Comment - Please answer the following questions for serviceability approval:

1. UFO -- does the UFO identify the most likely problems customers will see and identify how the feature will enable them to diagnose and solve those problems without resorting to raising a PMR? Have these issues been addressed in the implementation?

Yes, the UFO addresses problems customers might encounter. Warning messages were implemented and provides information about issues customers might encounter.

2. Test and Demo -- As part of the serviceability process we're asking feature teams to test and analyze common problem paths for serviceability and demo those problem paths to someone not involved in the development of the feature (eg. L2, test team, or another development team).
a) What problem paths were tested and demonstrated?
- Invalid `rolloverStartTime` and `rolloverInterval` values (to see the warnings)
- enable/disable time-based rollover

b) Who did you demo to?
**Tanya Kulik and Steven Schader**

c) Do the people you demo'd to agree that the serviceability of the demonstrated problem scenarios is sufficient to avoid PMRs for any problems customers are likely to encounter, or that L2 should be able to quickly address those problems without need to engage L3?
**Yes**

SVT -- SVT team is often the first team to try new features and often encounters problems setting up and using them. Note that we're not expecting SVT to do full serviceability testing -- just to sign-off on the serviceability of the problem paths they encountered.
a) Who conducted SVT tests for this feature?
**It was agreed no explicit SVT testing is needed for this feature, since the FAT coverage is enough.**
b) Do they agree that the serviceability of the problems they encountered is sufficient to avoid PMRs, or that L2 should be able to quickly address those problems without need to engage L3?
**Yes**

Which L2 / L3 queues will handle PMRs for this feature? Ensure they are present in the contact reference file and in the queue contact summary, and that the respective L2/L3 teams know they are supporting it. Ask Don Bourne if you need links or more info.
**WAS L3: Liberty Log Analytics**

Does this feature add any new metrics or emit any new JSON events? If yes, have you updated the JMX metrics reference list / Metrics reference list / JSON log events reference list in the Open Liberty docs?
**No**

@donbourne, for your review above. Thanks!

**Comment by chirp1** [comment]
Information to document is in the issue at https://github.com/OpenLiberty/docs/issues/5637 . Apptoving.

**Comment by ellen-lau** [comment]
@cbridgha can I get externals approval for this? there are no API changes, just additions to configuration.

**Comment by cbridgha** [comment]
Externals approved - For any config changes - we ensure the delivered metatype changes match what was explained in the UFO review.
